synapseclient/core/download/download_functions.py

```diff
@@ -0,0 +1,727 @@
+        IOError:                 If the local file does not exist.
+        SynapseError:            If fail to download the file.
+        SynapseHTTPError:        If there are too many redirects.
+        SynapseMd5MismatchError: If the actual MD5 does not match expected MD5.
+
+    Returns:
+        The path to downloaded file or None if the download failed
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+
+    destination = os.path.abspath(destination)
+    actual_md5 = None
+    redirect_count = 0
+    delete_on_md5_mismatch = True
+    client.logger.debug(f"Downloading from {url} to {destination}")
+    while redirect_count < REDIRECT_LIMIT:
+        redirect_count += 1
+        scheme = urllib_urlparse.urlparse(url).scheme
+        if scheme == "file":
+            delete_on_md5_mismatch = False
+            destination = utils.file_url_to_path(url, verify_exists=True)
+            if destination is None:
+                raise IOError(f"Local file ({url}) does not exist.")
+            break
+        elif scheme == "sftp":
+            username, password = client._getUserCredentials(url)
+            destination = SFTPWrapper.download_file(
+                url=url,
+                localFilepath=destination,
+                username=username,
+                password=password,
+                show_progress=not client.silent,
+            )
+            break
+        elif scheme == "ftp":
+            transfer_start_time = time.time()
+
+            def _ftp_report_hook(
+                block_number: int,
+                read_size: int,
+                total_size: int,
+                destination: str = destination,
+                transfer_start_time: float = transfer_start_time,
+            ) -> None:
+                """Report hook for urllib.request.urlretrieve to show download progress.
+
+                Arguments:
+                    block_number: The number of blocks transferred so far
+                    read_size: The size of each block
+                    total_size: The total size of the file
+                    destination: The destination file path
+                    transfer_start_time: The time when the transfer started
+
+                Returns:
+                    None
+                """
+                show_progress = not client.silent
+                if show_progress:
+                    client._print_transfer_progress(
+                        transferred=block_number * read_size,
+                        toBeTransferred=total_size,
+                        prefix="Downloading ",
+                        postfix=os.path.basename(destination),
+                        dt=time.time() - transfer_start_time,
+                    )
+
+            urllib_request.urlretrieve(
+                url=url, filename=destination, reporthook=_ftp_report_hook
+            )
+            break
+        elif scheme == "http" or scheme == "https":
```

Comment:
nit: easier to read: `elif scheme in ["http", "https"]`

Reply:
Great suggestion!

Reply:
Added in https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1106/commits/9e8a1226cafca81d3fb0956ca552322e2842c71d